pkg/ssh/ssh_keys.go

```diff
@@ -0,0 +1,137 @@
+package ssh
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+
+	"github.com/cli/cli/v2/internal/config"
+	"github.com/cli/cli/v2/internal/run"
+	"github.com/cli/safeexec"
+)
+
+type SshContext struct {
+	ConfigDir string
+	KeygenExe string
+}
+
+type SshKeyPair struct {
+	PublicKeyPath  string
+	PrivateKeyPath string
+}
+
+func (c *SshContext) LocalPublicKeys() ([]string, error) {
+	sshDir, err := c.sshDir()
+	if err != nil {
+		return nil, err
+	}
+
+	return filepath.Glob(filepath.Join(sshDir, "*.pub"))
+}
+
+func (c *SshContext) HasKeygen() bool {
+	_, err := c.findKeygen()
+	return err == nil
+}
+
+type GenerateSSHKeyOptions struct {
+	noErrorOnExists  bool
+	promptPassphrase func() (string, error)
+}
+
+func WithNoErrorOnExistingKey() func(*GenerateSSHKeyOptions) {
```

Comment:
Instead of having an option that control error behavior, why not just return a specific error type that signals that a key already exists? Then the caller can just handle that error type and keep using the SshKeyPair value while discarding the error.

Reply:
I had considered that, and prefer that as well, but I wasn't sure if it was OK in Go to return and error and also a valid value. Sounds like it is in which case I'll switch to that